### PR TITLE
Implement non_fast_start optimization in MP4 ISOM demuxer.

### PR DESCRIPTION
**Pull request by varsill**

### CODE REVIEW

File: lib/membrane_mp4/demuxer/isom.ex [hunk 1]
```diff
@@ -126,28 +199,141 @@ defmodule Membrane.MP4.Demuxer.ISOM do
...
+    end
+  end
+
+  defp handle_non_fast_start_optimization(%{fsm_state: :skip_mdat} = state) do
+    box_after_mdat_beginning = state.mdat_beginning + @header_size + state.mdat_size
+    seek(state, box_after_mdat_beginning, :infinity, false)
+  end
+
+  defp handle_non_fast_start_optimization(%{fsm_state: :go_back_to_mdat} = state) do
+    seek(state, state.mdat_beginning, state.mdat_size + @header_size, true)
```
Comment (mat-hek) [inline lib/membrane_mp4/demuxer/isom.ex:308]:
```suggestion
    seek(state, state.mdat_beginning, state.mdat_size + @header_size, false)
```

File: lib/membrane_mp4/demuxer/isom.ex [hunk 2]
```diff
@@ -126,28 +199,141 @@ defmodule Membrane.MP4.Demuxer.ISOM do
...
+          %Membrane.File.SeekSourceEvent{start: start, size_to_read: size_to_read, last?: last?}},
+       demand: :input
+     ], state}
+  end
+
   defp handle_can_read_mdat_box(ctx, state) do
-    state = %{state | samples_info: SamplesInfo.get_samples_info(state.boxes[:moov])}
+    {seek_events, state} =
+      if state.optimize_for_non_fast_start? do
+        seek(state, :cur, :infinity, true)
```
Comment (mat-hek) [inline lib/membrane_mp4/demuxer/isom.ex:329]: let's add a comment for this

File: lib/membrane_mp4/demuxer/isom.ex [hunk 3]
```diff
@@ -126,28 +199,141 @@ defmodule Membrane.MP4.Demuxer.ISOM do
...
+            else: state
+
+        handle_non_fast_start_optimization(state)
+
+      true ->
+        {[demand: :input], state}
+    end
+  end
+
+  defp update_fsm_state(state) do
```
Comment (mat-hek) [inline lib/membrane_mp4/demuxer/isom.ex:243]: as we discussed, add the context argument

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane_mp4/demuxer/isom.ex [hunk 4]
```diff
@@ -34,41 +53,72 @@ defmodule Membrane.MP4.Demuxer.ISOM do
...
+      {[demand: :input], state}
+    end
   end
 
   @impl true
-  def handle_stream_format(:input, _stream_format, _ctx, state) do
-    {[], state}
+  def handle_event(
+        :input,
+        %Membrane.File.NewSeekEvent{},
```
[the remark this answers is not in the document]
Comment (varsill) [reply lib/membrane_mp4/demuxer/isom.ex:90]: done

File: lib/membrane_mp4/demuxer/isom.ex (hunk 1 above)
Comment (varsill) [reply lib/membrane_mp4/demuxer/isom.ex:308]: done

File: lib/membrane_mp4/demuxer/isom.ex (hunk 2 above)
Comment (varsill) [reply lib/membrane_mp4/demuxer/isom.ex:329]: done

File: lib/membrane_mp4/demuxer/isom.ex [hunk 5]
```diff
@@ -89,14 +139,37 @@ defmodule Membrane.MP4.Demuxer.ISOM do
...
+
+  def handle_process(
+        :input,
+        _buffer,
+        _ctx,
+        %{
+          fsm_state: fsm_state
+        } = state
+      )
+      when fsm_state in [:mdat_skipping, :going_back_to_mdat] do
```
[the remark this answers is not in the document]
Comment (varsill) [reply lib/membrane_mp4/demuxer/isom.ex]: I have improved it, done

File: lib/membrane_mp4/demuxer/isom.ex (hunk 3 above)
Comment (varsill) [reply lib/membrane_mp4/demuxer/isom.ex]: done

**Review by mat-hek — APPROVED** [review]

File: lib/membrane_mp4/demuxer/isom.ex [hunk 6]
```diff
@@ -126,28 +178,149 @@ defmodule Membrane.MP4.Demuxer.ISOM do
...
+            else: state
+
+        handle_non_fast_start_optimization(state)
+
+      true ->
+        {[demand: :input], state}
+    end
+  end
+
+  defp update_fsm_state(state, ctx \\ []) do
```
[the remark this answers is not in the document]
Comment (varsill) [reply lib/membrane_mp4/demuxer/isom.ex]: done

**Comment by varsill** [comment]
We need to wait until https://github.com/membraneframework/membrane_file_plugin/pull/40 is merged and new version of membrane_file_plugin is released, so that we can change the deps in this PR

**Pull request merged**